File: app/src/main/java/com/dpis/module/DpisApplication.java
package com.dpis.module;

import com.dpis.module.root.RootAccessProbe;
import com.dpis.module.fonts.HyperOsNativeProxyAssetExporter;
import com.dpis.module.updates.UpdatePackageInstaller;

import android.app.Application;
import android.content.Context;
import com.google.android.material.color.DynamicColors;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import io.github.libxposed.service.XposedService;
import io.github.libxposed.service.XposedServiceHelper;

public final class DpisApplication extends Application implements XposedServiceHelper.OnServiceListener {
    private static final long UPDATE_CACHE_STARTUP_MAX_AGE_MS = 24 * 60 * 60 * 1000L;

    interface ServiceStateListener {
        void onServiceStateChanged();
    }

    private static final Set<ServiceStateListener> SERVICE_STATE_LISTENERS =
            new CopyOnWriteArraySet<>();

    private static volatile DpisApplication instance;
    private static volatile DpisConfigStore configStore;
    private static volatile XposedService xposedService;
    private static volatile boolean xposedSelfLoaded;
    public boolean xposedSelfLoadedByLegacyConstructorHook;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        DpisLog.setAppLogSink(new DpisAppLogStore(this));
        DpisLog.i("app process started");
        RootAccessProbe.warmUpAsync();
        DynamicColors.applyToActivitiesIfAvailable(this);
        HyperOsNativeProxyAssetExporter.exportBundledNativeProxyLibrary(this, DpisLog::e);
        configStore = ConfigStoreFactory.createLocalModuleConfigStore(this);
        migrateLocalConfigStore(configStore);
        DpisLog.setLoggingEnabled(configStore.isGlobalLogEnabled());
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(configStore);
        XposedServiceHelper.registerListener(this);
        UpdatePackageInstaller.clearStaleUpdateCache(this, UPDATE_CACHE_STARTUP_MAX_AGE_MS);
    }

    @Override
    public void onServiceBind(XposedService service) {
        DpisConfigStore localStore = ConfigStoreFactory.createLocalModuleConfigStore(this);
        DpisConfigStore runtimeDeliveryStore =
                ConfigStoreFactory.createRuntimeDeliveryModuleConfigStore(service);
        migrateLocalConfigStore(localStore);
        migrateLocalConfigStore(runtimeDeliveryStore);
        configStore = ConfigStoreFactory.createLocalUiModuleConfigStore(this, service);
        DpisLog.setLoggingEnabled(configStore.isGlobalLogEnabled());
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(configStore);
        xposedService = service;
        notifyServiceStateChanged();
    }

    @Override
    public void onServiceDied(XposedService service) {
        configStore = ConfigStoreFactory.createLocalModuleConfigStore(this);
        DpisLog.setLoggingEnabled(configStore.isGlobalLogEnabled());
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(configStore);
        xposedService = null;
        notifyServiceStateChanged();
    }

    static DpisConfigStore getConfigStore() {
        return configStore;
    }

    public static XposedService getXposedService() {
        return xposedService;
    }

    static void markXposedSelfLoaded() {
        xposedSelfLoaded = true;
        notifyServiceStateChanged();
    }

    static boolean isXposedSelfLoaded() {
        DpisApplication application = instance;
        return xposedSelfLoaded
                || (application != null && application.xposedSelfLoadedByLegacyConstructorHook);
    }

    static void clearXposedSelfLoadedForTest() {
        xposedSelfLoaded = false;
        DpisApplication application = instance;
        if (application != null) {
            application.xposedSelfLoadedByLegacyConstructorHook = false;
        }
    }

    public static DpisConfigStore getActiveHookConfigStore(Context context) {
        DpisConfigStore sharedStore = configStore;
        if (sharedStore != null) {
            return sharedStore;
        }
        if (context == null) {
            return null;
        }
        XposedService service = xposedService;
        return service != null
                ? ConfigStoreFactory.createLocalUiModuleConfigStore(context, service)
                : ConfigStoreFactory.createLocalModuleConfigStore(context);
    }

    static void reloadConfigStore() {
        DpisApplication application = instance;
        if (application == null) {
            return;
        }
        XposedService service = xposedService;
        DpisConfigStore localStore = ConfigStoreFactory.createLocalModuleConfigStore(application);
        migrateLocalConfigStore(localStore);
        migrateLocalConfigStore(ConfigStoreFactory.createRuntimeDeliveryModuleConfigStore(service));
        DpisConfigStore refreshedStore = service != null
                ? ConfigStoreFactory.createLocalUiModuleConfigStore(application, service)
                : ConfigStoreFactory.createLocalModuleConfigStore(application);
        configStore = refreshedStore;
        DpisLog.setLoggingEnabled(refreshedStore.isGlobalLogEnabled());
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(refreshedStore);
        notifyServiceStateChanged();
    }

    static void addServiceStateListener(ServiceStateListener listener, boolean notifyImmediately) {
        SERVICE_STATE_LISTENERS.add(listener);
        if (notifyImmediately) {
            listener.onServiceStateChanged();
        }
    }

    static void removeServiceStateListener(ServiceStateListener listener) {
        SERVICE_STATE_LISTENERS.remove(listener);
    }

    private static void notifyServiceStateChanged() {
        for (ServiceStateListener listener : SERVICE_STATE_LISTENERS) {
            listener.onServiceStateChanged();
        }
    }

    private static void migrateLocalConfigStore(DpisConfigStore store) {
        if (store == null) {
            return;
        }
        store.migrateLegacyWechatDpi();
        store.migrateLegacyPackageConfigToAggregated();
    }

    private static void publishRuntimeConfig(DpisConfigStore from, DpisConfigStore to) {
        if (from == null || to == null || from == to) {
            return;
        }
        // LSPosed remote preferences are a runtime delivery copy, not a migration
        // source or backup. Publish only runtime-shared config from the local store.
        Map<String, Object> snapshot = from.snapshotRuntimeDelivery();
        to.replaceAll(snapshot);
    }
}

File: app/src/main/java/com/dpis/module/DpisApplication.kt
package com.dpis.module

import android.app.Application
import android.content.Context
import com.dpis.module.DpisLog.AppLogSink
import com.dpis.module.diagnostics.DpisAppLogStore
import com.dpis.module.fonts.HyperOsNativeProxyAssetExporter
import com.dpis.module.fonts.TypefaceCatalogCache.preload
import com.dpis.module.root.RootAccessProbe
import com.dpis.module.runtime.RuntimeConfigDelivery
import com.dpis.module.runtime.RuntimePropertyRecoveryCoordinator
import com.dpis.module.updates.UpdatePackageInstaller
import com.google.android.material.color.DynamicColors
import io.github.libxposed.service.XposedService
import io.github.libxposed.service.XposedServiceHelper
import io.github.libxposed.service.XposedServiceHelper.OnServiceListener
import java.util.concurrent.CopyOnWriteArraySet
import kotlin.concurrent.Volatile

class DpisApplication : Application(), OnServiceListener {
    interface ServiceStateListener {
        fun onServiceStateChanged()
    }

    var xposedSelfLoadedByLegacyConstructorHook: Boolean = false

    override fun onCreate() {
        super.onCreate()
        instance = this
        val appLogStore = DpisAppLogStore(this)
        DpisLog.setAppLogSink(AppLogSink { level: String?, message: String? ->
            appLogStore.record(
                level,
                message
            )
        })
        DpisLog.i("app process started")
        RootAccessProbe.warmUpAsync()
        DynamicColors.applyToActivitiesIfAvailable(this)
        HyperOsNativeProxyAssetExporter.exportBundledNativeProxyLibrary(
            this,
            HyperOsNativeProxyAssetExporter.Logger { msg: String?, throwable: Throwable? ->
                DpisLog.e(
                    msg,
                    throwable
                )
            })
        // Migrate the private catalog before remote config mirroring can touch the old XML.
        ConfigStoreFactory.createLocalFontLibraryStore(this)
        preload(this)
        configStore = ConfigStoreFactory.createLocalModuleConfigStore(this)
        migrateLocalConfigStore(configStore)
        val initializedStore: DpisConfigStore? = configStore
        if (initializedStore == null) {
            DpisLog.e(
                "app config store initialization returned null",
                IllegalStateException("config store unavailable")
            )
        } else {
            DpisLog.setLoggingEnabled(initializedStore.isGlobalLogEnabled())
            RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(initializedStore)
        }
        XposedServiceHelper.registerListener(this)
        UpdatePackageInstaller.clearStaleUpdateCache(this, UPDATE_CACHE_STARTUP_MAX_AGE_MS)
    }

    override fun onServiceBind(service: XposedService) {
        val localStore = ConfigStoreFactory.createLocalModuleConfigStore(this)
        val runtimeDeliveryStore =
            ConfigStoreFactory.createRuntimeDeliveryModuleConfigStore(service)
        migrateLocalConfigStore(localStore)
        migrateLocalConfigStore(runtimeDeliveryStore)
        configStore = ConfigStoreFactory.createLocalUiModuleConfigStore(this, service)
        DpisLog.setLoggingEnabled(configStore!!.isGlobalLogEnabled())
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(configStore)
        xposedService = service
        notifyServiceStateChanged()
    }

    override fun onServiceDied(service: XposedService) {
        configStore = ConfigStoreFactory.createLocalModuleConfigStore(this)
        DpisLog.setLoggingEnabled(configStore!!.isGlobalLogEnabled())
        RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(configStore)
        xposedService = null
        notifyServiceStateChanged()
    }

    companion object {
        private val UPDATE_CACHE_STARTUP_MAX_AGE_MS = 24 * 60 * 60 * 1000L

        init {
            RuntimeConfigDelivery.setLocalSnapshotReloader(Runnable { reloadConfigStore() })
        }

        private val SERVICE_STATE_LISTENERS: MutableSet<ServiceStateListener> =
            CopyOnWriteArraySet<ServiceStateListener>()

        @Volatile
        private var instance: DpisApplication? = null

        @Volatile
        private var configStore: DpisConfigStore? = null

        @Volatile
        @JvmStatic
        var xposedService: XposedService? = null
            private set

        @Volatile
        private var xposedSelfLoaded = false

        @JvmStatic
        fun getConfigStore(): DpisConfigStore {
            return configStore!!
        }

        fun markXposedSelfLoaded() {
            xposedSelfLoaded = true
            notifyServiceStateChanged()
        }

        @JvmStatic
        fun isXposedSelfLoaded(): Boolean {
            val application: DpisApplication? = instance
            return xposedSelfLoaded
                    || (application != null && application.xposedSelfLoadedByLegacyConstructorHook)
        }

        @JvmStatic
        fun clearXposedSelfLoadedForTest() {
            xposedSelfLoaded = false
            val application: DpisApplication? = instance
            if (application != null) {
                application.xposedSelfLoadedByLegacyConstructorHook = false
            }
        }

        @JvmStatic
        fun getActiveHookConfigStore(context: Context?): DpisConfigStore? {
            val sharedStore: DpisConfigStore? = configStore
            if (sharedStore != null) {
                return sharedStore
            }
            if (context == null) {
                return null
            }
            val service: XposedService? = xposedService
            return if (service != null)
                ConfigStoreFactory.createLocalUiModuleConfigStore(context, service)
            else
                ConfigStoreFactory.createLocalModuleConfigStore(context)
        }

        fun reloadConfigStore() {
            val application: DpisApplication? = instance
            if (application == null) {
                return
            }
            val service: XposedService? = xposedService
            val localStore = ConfigStoreFactory.createLocalModuleConfigStore(application)
            migrateLocalConfigStore(localStore)
            migrateLocalConfigStore(
                ConfigStoreFactory.createRuntimeDeliveryModuleConfigStore(
                    service
                )
            )
            val refreshedStore = if (service != null)
                ConfigStoreFactory.createLocalUiModuleConfigStore(application, service)
            else
                ConfigStoreFactory.createLocalModuleConfigStore(application)
            configStore = refreshedStore
            DpisLog.setLoggingEnabled(refreshedStore.isGlobalLogEnabled())
            RuntimePropertyRecoveryCoordinator.resyncConfiguredTargetsAsync(refreshedStore)
            notifyServiceStateChanged()
        }

        @JvmStatic
        fun addServiceStateListener(listener: ServiceStateListener, notifyImmediately: Boolean) {
            SERVICE_STATE_LISTENERS.add(listener)
            if (notifyImmediately) {
                listener.onServiceStateChanged()
            }
        }

        @JvmStatic
        fun removeServiceStateListener(listener: ServiceStateListener?) {
            SERVICE_STATE_LISTENERS.remove(listener)
        }

        private fun notifyServiceStateChanged() {
            for (listener in SERVICE_STATE_LISTENERS) {
                listener.onServiceStateChanged()
            }
        }

        private fun migrateLocalConfigStore(store: DpisConfigStore?) {
            if (store == null) {
                return
            }
            store.migrateLegacyWechatDpi()
            store.migrateLegacyPackageConfigToAggregated()
        }

        @JvmStatic
        private fun publishRuntimeConfig(from: DpisConfigStore?, to: DpisConfigStore?) {
            if (from == null || to == null || from == to) {
                return
            }
            // LSPosed remote preferences are a runtime delivery copy, not a migration
            // source or backup. Publish only runtime-shared config from the local store.
            val snapshot = from.snapshotRuntimeDelivery()
            to.replaceAll(snapshot)
        }
    }
}
